hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java

```diff
@@ -94,9 +105,84 @@ public static JoinType toHibernateJoinType(int astJoinType) {
 		}
 	}
 
+	private <T extends AST> List<T> findAllNodes(AST node, Class<T> clazz) {
+		ArrayList<T> result = new ArrayList<>();
+		findAllNodesCore( node, clazz, result );
+		return result;
+	}
+
+	private <T extends AST> void findAllNodesCore(AST node, Class<T> clazz, List<T> found) {
+		if ( clazz.isAssignableFrom( node.getClass() ) ) {
+			found.add( (T) node );
+		}
+		if ( node.getFirstChild() != null ) {
+			findAllNodesCore( node.getFirstChild(), clazz, found );
+		}
+		if ( node.getNextSibling() != null ) {
+			findAllNodesCore( node.getNextSibling(), clazz, found );
+		}
+	}
+
+	private Set<String> findQueryReferencedTables(QueryNode query) {
+		if ( !walker.getSessionFactoryHelper()
+				.getFactory()
+				.getSessionFactoryOptions()
+				.isOmitJoinOfSuperclassTablesEnabled() ) {
+			if (LOG.isDebugEnabled()) {
+				LOG.debug( "Finding of query referenced tables is skipped because the feature is disabled by " +
+							"configuration. See " + AvailableSettings.OMIT_JOIN_OF_SUPERCLASS_TABLES );
+			}
+			return null;
+		}
+
+		if ( walker.getEnabledFilters() != null && !walker.getEnabledFilters().isEmpty() ) {
+			LOG.debug( "Finding of query referenced tables is skipped because filters are enabled." );
+			return null;
+		}
+
+		if ( LOG.isDebugEnabled() ) {
+			LOG.debug( TokenPrinters.REFERENCED_TABLES_PRINTER.showAsString(
+					query,
+					"Tables referenced from query nodes:"
+			) );
+		}
+
+		Set<String> result = new HashSet<>();
+
+		// Find tables referenced by FromReferenceNodes
+		List<FromReferenceNode> fromReferenceNodes = findAllNodes( query, FromReferenceNode.class );
+		for ( FromReferenceNode node : fromReferenceNodes ) {
+			String[] tables = node.getReferencedTables();
+			if ( tables != null ) {
+				Arrays.stream( tables )
+						.filter( Objects::nonNull )
+						.forEach( result::add );
+			}
+		}
+
+		// Find tables referenced by fromElementsForLoad
+		if ( query.getSelectClause() != null ) {
+			for ( Object element : query.getSelectClause().getFromElementsForLoad() ) {
+				FromElement fromElement = (FromElement) element;
+				EntityPersister entityPersister = fromElement.getEntityPersister();
+				if ( entityPersister != null && entityPersister instanceof AbstractEntityPersister ) {
+					AbstractEntityPersister aep = (AbstractEntityPersister) entityPersister;
```

Comment:
Maybe it's a stupid question but what if `canOmitSuperclassTableJoin()` of this persister returns false?

Reply:
It does not matter. This method looks for all tables which are referenced by the query. And persister here just returns all tables which are used for entity mapping because all these tables are referenced.

`canOmitSuperclassTableJoin()` is used later on when a persister creates SQL from part of the query. Then a persister, if the `canOmitSuperclassTableJoin()` is `true`, does not have to use all tables bud can omit some of them.